src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -182,14 +185,37 @@ public void beforeCommand(CommandEnvironment env) throws AbruptExitException {
         env.getOutputBase().getRelative(env.getRuntime().getProductName() + "-remote-logs");
     cleanAndCreateRemoteLogsDir(logDir);
 
+    if (Strings.isNullOrEmpty(remoteOptions.remoteCache) && remoteOptions.remoteCacheHeaders.size() > 0){
+      throw new AbruptExitException(
+          "--remote_cache_header can only be used if --remote_cache is set.",
+          ExitCode.COMMAND_LINE_ERROR);
+    }
+    if (Strings.isNullOrEmpty(remoteOptions.remoteExecutor) && remoteOptions.remoteCacheHeaders.size() > 0){
+      throw new AbruptExitException(
+          "--remote_exec_header can only be used if --remote_executor is set.",
+          ExitCode.COMMAND_LINE_ERROR);
+    }
     try {
       List<ClientInterceptor> interceptors = new ArrayList<>();
+      List<ClientInterceptor> execInterceptors = interceptors;
+      List<ClientInterceptor> cacheInterceptors = interceptors;
       if (remoteOptions.experimentalRemoteGrpcLog != null) {
         rpcLogFile =
             new AsynchronousFileOutputStream(
                 env.getWorkingDirectory().getRelative(remoteOptions.experimentalRemoteGrpcLog));
         interceptors.add(new LoggingInterceptor(rpcLogFile, env.getRuntime().getClock()));
       }
+      if (remoteOptions.remoteHeaders.size() > 0){
+        interceptors.add(createExtraHeadersClientInterceptor(remoteOptions.remoteHeaders));
+      }
+      if (remoteOptions.remoteExecHeaders.size() > 0) {
+        execInterceptors = new ArrayList<>(interceptors);
+        execInterceptors.add(createExtraHeadersClientInterceptor(remoteOptions.remoteExecHeaders));
+      }
+      if (remoteOptions.remoteCacheHeaders.size() > 0) {
+        cacheInterceptors = new ArrayList<>(interceptors);
+        cacheInterceptors.add(createExtraHeadersClientInterceptor(remoteOptions.remoteCacheHeaders));
```

Comment:
Instead of adding the headers in the module, consider passing the headers to the RemoteSpawnRunner. That way it's easier to write a unit test.